### PR TITLE
Make any callable compatible with (*args: Any, **kwargs: Any)

### LINKED ISSUE #5876: Make any callable compatible with (*args: Any, **kwargs: Any)?
Perhaps a callable that accepts arbitrary arguments by having `(*args: Any, **kwargs: Any)` could be treated similarly to `Callable[..., X]` (with explicit `...`), i.e. a callable with arbitrary arguments would be compatible with it.

The reason for this is that there's no easy way to annotate a function so that the type would be `Callable[..., X]`, even though this seems to be a somewhat common need. Users sometimes assume that using `(*args: Any, **kwargs: Any)` will work for this purpose and are confused when it doesn't. https://github.com/python/typing/issues/264#issuecomment-436525286 is a recent example.

Example:

```py
from typing import Any

class A:
    # A method that accepts unspecified arguments and returns int
    def f(self, *args: Any, **kwargs: Any) -> int: pass

class B(A):
    def f(self, x: int) -> int: pass  # Should be okay
```

We could perhaps extend this to handle decorators as well:

```py
from typing import Any, Callable, TypeVar

T = TypeVar('T', bound=Callable[..., Any])

def deco(f: T) -> T:
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        print('called')
        return f(*args, **kwargs)
    return wrapper  # Maybe this should be okay
```

### PR DESCRIPTION
**Pull request by hauntsaninja**
Resolves #5876

### CODE REVIEW

File: test-data/unit/check-classes.test [hunk 1]
```diff
@@ -6891,3 +6891,14 @@ class B(A):  # E: Final class __main__.B has abstract attributes "foo"
...
+
+
+[case testArgsKwargsInheritance]
+from typing import Any
+
+class A(object):
+    def f(self, *args: Any, **kwargs: Any) -> int: ...
+
+class B(A):
+    def f(self, x: int) -> int: ...
```
Comment (sobolevn) [inline test-data/unit/check-classes.test:6903]: I understand that this is very handy in some cases, https://github.com/python/mypy/issues/5876 has a handful of them. But, it is important to remember that this is not type-safe.

```python
from typing import Any

class A(object):
    def f(self, *args: Any, **kwargs: Any) -> int: ...

class B(A):
    def f(self, x: int) -> int: ...

def test(arg: A):
    arg.f(x=0, kw1=1, kw2=2)

test(A())  # ok
test(B())  # mypy is ok with that, but raises TypeError in real-life
```

It breaks Liskov's law.

Maybe we can somehow limit the scope of this new rule? For example:
- It is fine for decorator example from https://github.com/python/mypy/issues/5876
- It is fine for `Protocol` check (when protocol defines less-or-equally broad signature): `Procol.method(self, arg: int)` -> is ok with `Impl.method(self, *args, **kwargs)`
- It is fine for callbacks
- Maybe something else
- It is **not** fine for subclassing

Comment (hauntsaninja) [reply test-data/unit/check-classes.test:6903]: Yeah, I think the ask is explicitly for something that is not type safe. The check when "protocol defines less-or-equally broad signature" should already work. The decorator example from #5876 is a little interesting; this PR doesn't actually change inference there.

#5876 is probably a top five most popular mypy issue, so ideally something would be done. Maybe the answer is another flag.

**Review by ilevkivskyi — APPROVED** [review]
FWIW I think we should just go ahead with this.

**Comment by ilevkivskyi** [comment]
It looks like `mypy_primer` tells us we are on right track :-)

**Comment by vikigenius** [comment]
@ilevkivskyi 

Obviously I am happy that this issue was resolved, since it's such a popular request but it bothers me that such an obvious violation of LSP is now apparently ok.

The primary use cases for using something like this in subclassing comes from the pytorch world

where the base class module might define a function like this


```python
class Module(object):
    def forward(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any]]:
        raise NotImplementedError
```

What the above means is that it's just a contract, subclasses should technically be able to call forward with any arguments. 

```python
class SummarizationModule(Module)
    def forward(self, src: torch.Tensor, tgt: torch.Tensor) -> Tensor:
        return torch.zeros(2, 3)
```

It does not violate LSP technically even if it does because you will never call the base class forward anyway since it is not implemented. In compiled languages this won't be a problem since the base forward acts like a pure virtual function and the language does not even allow you to call it.

But in the general case if the forward is not performing the role of a pure virtual function then allowing subclasses to call specific arguments is an obvious violation of LSP.

Take this obviously contrived example

```python
class Base(object):
    def count(self, *args, **kwargs):
        return len(args) + len(kwargs)

class Derived(Base):
    def count(self, a: int): # This should never be allowed since it's not polymorphic, just use a different name
        return a
```

Right now python afaik gives us no mechanism to distinguish these two cases, and I can understand why this was accepted, but it is stil problematic.

**Comment by hauntsaninja** [comment]
I think cases like yours could be annotated as `*args: object, **kwargs: object` to get LSP errors again.

**Comment by ilevkivskyi** [comment]
Yeah, I agree with @hauntsaninja just use `object` instead of `Any`.

**Pull request merged**